Add CI pipeline to generate a new beta

mokagio opened the pull request:
### Fix
Adds the CI pipeline required to trigger the new beta deployment from CI and updates the `new_beta_release` lane to also create the backmerge PR.

To get there, I ended up updating a lot more code, including:

- Upgrading to `release-toolkit` 12.0
- Consolidating to `create_release_backmerge_pull_request`
- Remove some unused lanes
- Move more lanes to dedicated files and adopt named block parameters

See also paaHJt-6Zx-p2 paaHJt-6ps-p2

### Test
This will be tested properly only once landed on the release branch.

mokagio committed:
Add CI pipeline to generate a new beta

mokagio committed:
Create release backmerge as part of `new_beta_release`

mokagio committed:
Create backmerge PR for new beta

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/simplenote-ios/milestone/69">4.54</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Use named block parameters in `new_hotfix_release` lane

mokagio committed:
Use named block parameters in `finalize_hotfix_release` lane

mokagio committed:
Use named block parameters in `build_and_upload_release` lane

mokagio committed:
Use named block parameters in `build_and_upload_internal` lane

mokagio committed:
Use named block parameters in `update_metadata_on_app_store_connect` lane

mokagio committed:
Remove deprecated `ios_*` actions from `new_hotfix_release`

mokagio committed:
Move `new_hotfix_release` to `release.rb`

mokagio committed:
Move `finalize_hotfix_release` to `release.rb`

mokagio committed:
Remove deprecated `ios_*` actions from `finalize_hotfix_release`

mokagio committed:
Remove unused `build_and_upload_beta|stable_release|release` lanes

mokagio committed:
Delete unused `build_and_upload_internal` lane

mokagio committed:
Remove deprecated `ios_get_app_version` from update ASC metadata lane

mokagio committed:
Remove deprecated `ios_current_branch_is_hotfix` from `finalize_release`

mokagio committed:
Upgrade to release-toolkit 12.0

mokagio committed:
Consolidate backmerge PR creation via single `release-toolkit` action

mokagio committed:
Add lane to trigger new beta release in CI

mokagio commented on fastlane/Fastfile:
```diff
@@ -77,206 +77,20 @@ ENV['PUBLIC_CONFIG_FILE'] = VERSION_FILE_PATH
-    ios_finalize_prechecks(options)
-    version = ios_get_app_version(public_version_xcconfig_file: VERSION_FILE_PATH)
-    trigger_release_build(branch_to_build: "release/#{version}")
-  end
```
👆  `new_hotfix_release` and `finalize_hotfix_release` have been moved to `release.rb`.

👇 the other `build_and_upload_...` lanes below were unused and have been removed.

mokagio commented on fastlane/Fastfile:
```diff
@@ -77,206 +77,20 @@ ENV['PUBLIC_CONFIG_FILE'] = VERSION_FILE_PATH
     upload_to_app_store(
       app_identifier: APP_STORE_BUNDLE_IDENTIFIER,
-      app_version: ios_get_app_version(public_version_xcconfig_file: VERSION_FILE_PATH),
+      app_version: release_version_current,
```
This was required because of the `release-toolkit` 12.0 upgrade.

In hindsight, the upgrade was not required because `create_release_backmerge_pull_request` was already available in version 11, but it ought to have been done sooner or later, so might as well do it now.

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -61,7 +61,7 @@
 
   version = beta_release ? build_code_current : release_version_current
-  create_release(
+  create_github_release(
```
Another 12.0 upgrade requirement.

mokagio commented on fastlane/lanes/ci_releases.rb:
```diff
@@ -23,4 +23,15 @@
+      message: "New beta for #{release_version}",
+      environment: { RELEASE_VERSION: release_version }
+    )
+  end
```
So... this lane is not strictly necessary. After all, we use our internal release management app to trigger release builds.

I added it here because I plan to use it to test the new beta workflow once this PR lands as I don't want to reset the scenario for 4.53 in the release management app.

One could argue that it's useful to have these here as a fallback if the app is offline or something, but I think it might be overkill. In fact, maybe is best if I delete the whole file and use the Buildkite GUI to trigger this during testing.

What do you think?

WordPress has these lanes too, but that's because they were added before Rv2 was available. If we decide to remove the lanes from here, we shall delete them from there, too.

iangmaia replied:
Hmm I'm leaning towards deleting it.
I myself was confused for some time looking at the WPiOS automation, as we even mention it in the task in Rv2:
```
Run <code>bundle exec fastlane trigger_code_freeze_in_ci</code>.
```

While this lane is never really called, unless someone would do it directly (which [almost?] never happens these days).

mokagio replied:
Done via 24fd4a6471928a84e70af3d1db58f3d8f725b484. See also https://github.com/wordpress-mobile/WordPress-iOS/pull/23589

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -221,6 +224,77 @@
+  lane :new_hotfix_release do |version:, skip_confirm: false, skip_prechecks: false|
+    ensure_git_status_clean unless skip_prechecks
+
+    # Parse the provided version into an AppVersion object
```
```suggestion
```

mokagio committed:
Remove noisy comment

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -221,6 +224,76 @@
+
+    create_backmerge_prs
+
+    # Close hotfix milestone
```
```suggestion
```

mokagio committed:
Remove noisy comment

mokagio committed:
Remove var assignment that became unnecessary along the way

iangmaia commented on fastlane/lanes/release.rb:
```diff
@@ -262,18 +332,40 @@ def trigger_buildkite_release_build(branch:, beta:)
+def create_backmerge_pr!
+  pr_urls = create_backmerge_prs
+
+  return pr_urls unless pr_urls > 1
```
I haven't tested, but don't you need to compare the size?
```suggestion
  return pr_urls unless pr_urls.length > 1
```

mokagio replied:
Woops! 😳 

Thanks @iangmaia

mokagio replied:
For the record: I updated @iangmaia original suggestion from `size` to `length` and applied it from GitHub to maintain attribution.

Rationale: [`size` is an alias for `length`](https://ruby-doc.org/3.3.5/Array.html#method-i-size) and as far as I've seen, `length` is the most commonly used.

iangmaia commented on fastlane/lanes/release.rb:
```diff
@@ -262,18 +332,40 @@ def trigger_buildkite_release_build(branch:, beta:)
 end
 
-def create_release_backmerge_pr(version_to_merge:, next_version:)
+def create_backmerge_pr!
```
Given an error is thrown in an alternative flow (isn't guaranteed to always happen), does it still make sense to add a `!` to the method name?

iangmaia commented on fastlane/lanes/release.rb:
```diff
@@ -262,18 +332,40 @@ def trigger_buildkite_release_build(branch:, beta:)
+
+# Notice the plural in the name.
+# The action this method calls may create multiple backmerge PRs, depending on how many releas branches with version greater than the source are in the remote.
+def create_backmerge_prs
```
Interesting approach. I wonder if this is really safe to assume and, if so, if this approach should be ported to other apps.

AliSoftware replied:
@iangmaia [the `create_release_backmerge_pull_request` action from `release-toolkit` is already the one taking care of creating multiple PRs if there are multiple `release/*` branches with a greater version](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/lib/fastlane/plugin/wpmreleasetoolkit/actions/common/create_release_backmerge_pull_request_action.rb#L61-L73). So no need to port anything to other repos, it's already handled as long as you use the toolkit action.

iangmaia replied:
Yeah I know, but I meant **throwing an error when there are multiple (unexpected) PRs**, which is what @mokagio is doing differently here.

iangmaia commented on fastlane/lanes/release.rb:
````diff
@@ -262,18 +332,40 @@ def trigger_buildkite_release_build(branch:, beta:)
-    Please close any previous backmerge PR for `#{version_to_merge}`, delete the previous merge branch, then run the release task again.
+    Error creating backmerge pull request:
+
+    ```
+    #{e.message}
+    ```
````
```suggestion
    #{e.message}
```

One thing I realize happened is that the ``` does something strange, where the entire message (above and below) becomes part of this block. For now in new apps I'm working on (WPiOS/WPAndroid), I've just removed it and let only the line breaks, but perhaps we can think of a better formatting.

mokagio replied:
Removing the code fencing sounds good to me. Hopefully we'll see little of this error, but if it happens often and users found it poorly rendered, we can address then.

AliSoftware commented on fastlane/lanes/release.rb:
```diff
@@ -262,18 +332,40 @@ def trigger_buildkite_release_build(branch:, beta:)
+end
+
+# Notice the plural in the name.
+# The action this method calls may create multiple backmerge PRs, depending on how many releas branches with version greater than the source are in the remote.
```
```suggestion
# The action this method calls may create multiple backmerge PRs, depending on how many release branches with version greater than the source are in the remote.
```

mokagio replied:
Thanks @AliSoftware . I don't know what it is with the word "release" but I seem to have a high typo ratio for it 🙃

mokagio committed:
Fix typo: `s/releas/release`

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

mokagio committed:
Add required `.length` call in an array count check

Co-authored-by: Ian Guedes Maia <iangmaia@users.noreply.github.com>

mokagio committed:
Remove unused `ci_releases.rb` lanes

We no longer need repo-level automation to trigger release lanes.

See
https://github.com/Automattic/simplenote-ios/pull/1655#discussion_r1753900531

mokagio committed:
Remove code fencing from error message as it might mis-render

Co-authored-by: Ian Guedes Maia <iangmaia@users.noreply.github.com>

mokagio committed:
Remove import for removed lanes file

mokagio committed:
Rename `code_freeze` to `start_code_freeze` for consistency

mokagio commented:
I'm going to admin-merge this because I feel like the various comments have been addressed and there was nothing that was a "request changes". 

In particular, the backmerge chages will be useful in the upcoming 4.54 code freeze. As always, I'll be happy to follow up to address anything I might have missed.

mokagio commented:
😓 because I used an integration branch in #1659, the `release/4.53` branch was still around and that's where this PR landed 🤦‍♂️ 

I'll cherry-pick the commits from here into a new branch against `trunk` and admin-merge that...
